server/neptune/workflows/internal/terraform/root_locker.go

```diff
@@ -0,0 +1,72 @@
+package terraform
+
+import (
+	"context"
+	"errors"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/activities"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/root"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/terraform/state"
+	"go.temporal.io/sdk/workflow"
+)
+
+const (
+	DivergedStatus = "diverged"
+)
+
+type storeActivities interface {
+	FetchLatestDeployment(ctx context.Context, request activities.FetchLatestDeploymentRequest) (activities.FetchLatestDeploymentResponse, error)
+}
+
+type LockNotifier func(lockState *state.Lock) error
+
+type RootLocker struct {
+	Request  Request
+	Notifier LockNotifier
+	Ga       githubActivities
+	Sa       storeActivities
+}
+
+func (r *RootLocker) Lock(ctx workflow.Context) error {
+	// Fetch root's latest revision
+	var fetchLatestDeploymentResponse *activities.FetchLatestDeploymentResponse
+	err := workflow.ExecuteActivity(ctx, r.Sa.FetchLatestDeployment, activities.FetchLatestDeploymentRequest{
+		RepositoryName: r.Request.Repo.Name,
+		RootName:       r.Request.Root.Name,
+	}).Get(ctx, &fetchLatestDeploymentResponse)
+
+	if err != nil {
+		return err
+	}
+
+	// Compare with requested revision
+	var compareCommitsResponse *activities.CompareCommitsResponse
+	err = workflow.ExecuteActivity(ctx, r.Ga.CompareCommits, activities.CompareCommitsRequest{
+		Repo:      r.Request.Repo,
+		OldCommit: fetchLatestDeploymentResponse.Revision,
+		NewCommit: r.Request.Revision,
+	}).Get(ctx, &compareCommitsResponse)
+
+	if err != nil {
+		return err
+	}
```

Comment:
Why is this all happening in the terraform workflow?  It makes sense for this to be done in the parent since that's the one who's role is handling the deployment itself.   This workflow should ideally not know or care about deployments.